Standardize multisite admin screens and admin bar

felixarntz opened the pull request:
## Summary

This PR can be summarized in the following changelog entry:

* Standardize the multisite settings screen.

Once this is merged, the sub-issues that went into this must be considered as well (include each of their changelog entries and set the milestone accordingly):

* #9690 
* #9871 
* #9944 
* #9976 

## Relevant technical choices:

* See the respective sub-issues. All changes were already reviewed in one of those issues.

## Test instructions

This PR can be tested by following these steps:

* Before testing: All of those things have already been acceptance-tested through the sub-issues, so you don't need to be very thorough, as nothing was actually changed _in_ this PR. It's just an extra round to verify the following things work correctly.
* Set up a multisite, and activate the plugin network-wide.
* Go to the network admin.
* Go to the plugin's network settings screen.
* Change settings around in different ways, save them and verify they are persisted correctly.
* Go to the site admin.
* Go to the plugin's regular settings screen.
* Change some settings away from their defaults and save them.
* Go back to the network admin and the plugin's network settings screen.
* Restore the settings for the site where you just changed them.
* Go back to that site's plugin settings screen and ensure the settings are correctly set back to the defaults.
* Go to the network admin.
* Verify that the admin bar contains network-specific content, i.e. links to the network settings screen and sub-pages. It should not contain a notification bubble at the network level, since notifications are only relevant per site.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [x] I have added unittests to verify the code works as intended

Fixes #9447

afercia commented on js/src/wp-seo-network-admin.js:
```diff
@@ -0,0 +1,96 @@
+	 */
+	function handleAJAXSubmission( event ) {
+		var $form    = $( this );
+		var $submit  = $form.find( "[type='submit']:focus" );
```
Why getting the button (and thus its value to be use as action) only when focused? Browsers have slightly different behaviors when it comes to submitting forms using the keyboard. For example:
- go in the Restore Site tab
- with Chrome, I'm able to tab to the select element and focus it, press Enter and the form submits
- at that point, the button is not focused 
- the actual action is `yoast_handle_network_options` while it's supposed to be `yoast_restore_site`
- with Firefox:
- if I click the button using the mouse, the action is `yoast_handle_network_options`
- if I press the button using the keyboard, the action is `yoast_restore_site`

felixarntz replied:
It's necessary to find out through which button the form was submitted as that button determines the `action` to pass to the AJAX request (which on the server is then used as the AJAX action via `wp_ajax_{$action}`.

I adjusted the logic accordingly so that submitting without focusing a button still works:
* First, check if a button is focused. If so, use that.
* If not, find the current tab, and if there's a button inside that, use that.
* Otherwise fall back to the default action, which is for updating the options.

afercia commented on js/src/wp-seo-network-admin.js:
```diff
@@ -0,0 +1,96 @@
+
+				displaySettingsErrors( response.data );
+			},
+			error: function( xhr ) {
```
Shouldn't we favor the usage of `done()` and `fail()`?

afercia commented on js/src/wp-seo-network-admin.js:
```diff
@@ -0,0 +1,96 @@
+			prefix = wpseoNetworkAdminGlobalL10n.success_prefix;
+		}
+
+		a11ySpeak( prefix.replace( "%s", settingsErrors[ 0 ].message ), "assertive" );
```
❤️

afercia commented on js/src/wp-seo-network-admin.js:
```diff
@@ -0,0 +1,96 @@
+	 * @returns {void}
+	 */
+	function displaySettingsErrors( settingsErrors ) {
+		var $heading = $( ".wrap > h1" );
```
minor: this is used for both success and error messages, I'd consider to rename it

afercia commented on js/src/wp-seo-network-admin.js:
```diff
@@ -0,0 +1,96 @@
+	 *
+	 * @returns {void}
+	 */
+	function displaySettingsErrors( settingsErrors ) {
```
minor: this is used for both success and error messages, I'd consider to rename it

felixarntz replied:
I'd say we keep the name that way for parity with core. While not great there, the term "settings error" is used for all kinds of settings feedback messages, so the name here should indicate what it is responsible for.

afercia commented on admin/class-yoast-network-admin.php:
```diff
@@ -0,0 +1,299 @@
+		$whitelist_options = Yoast_Network_Settings_API::get()->get_whitelist_options( $option_group );
+
+		if ( empty( $whitelist_options ) ) {
+			add_settings_error( $option_group, 'settings_updated', __( 'You are not allowed to modify unregistered network settings.', 'wordpress-seo' ), 'error' );
```
Interesting: noticed now that `add_settings_error()` uses type `error` by default and then we use `updated`, I guess many other plugins do this. These are then used also as the CSS class names for the notices. However, `.updated` and `.error` are legacy in core. The new `notice-success` and `.notice-error` should be used instead. Should we open a ticket upstream?

felixarntz replied:
Agreed. I'd leave it as is in here until core decides to change its path regarding this. I opened https://core.trac.wordpress.org/ticket/44640

afercia commented on admin/class-yoast-network-admin.php:
```diff
@@ -0,0 +1,299 @@
+		$site = get_site( $site_id );
+		if ( ! $site ) {
+			/* translators: %s expands to the ID of a site within a multisite network. */
+			add_settings_error( $option_group, 'settings_updated', sprintf( __( 'Site %d not found.', 'wordpress-seo' ), $site_id ), 'error' );
```
Should we say, for example, `Site 4 not found` or `Site with ID 4 not found`?

afercia commented:
CR 🚧 
Reviewed a bit, mostly focusing on the JS and UI part. The PHP part would require some more expertise, not exactly my area.

A few considerations, seems to me the most important one is related to the form action that could fail depending on the browser and mouse/keyboard submission.

afercia commented:
One quick question: should the new built JS files be committed?
```
js/dist/wp-seo-network-admin-79-RC2.js
js/dist/wp-seo-network-admin-79-RC2.min.js
```

After running `grunt build:js` these are "untracked files".

afercia commented:
The test `Yoast_Network_Admin_Test::test_verify_request_with_invalid_nonce` passes on local (at least for me) but fails on Travis.

felixarntz commented:
I've rebuilt the new JS files so that they now carry the correct version number for trunk, and addressed @afercia's feedback.

moorscode commented on admin/class-yoast-network-admin.php:
```diff
@@ -0,0 +1,299 @@
+	public function handle_update_options_request() {
+		$option_group = filter_input( INPUT_POST, 'network_option_group', FILTER_SANITIZE_STRING );
+
+		$this->verify_request( "$option_group-network-options" );
```
When using a variable in a string like this, I'd really prefer to wrap it with braces, or concat the string.
`"${option_group}-network-options"`

felixarntz commented:
I fixed the merge conflict by moving the name of the new JS file over to the webpack `paths.js` file for the updated workflow, and addressed the concern by @moorscode.

jcomack commented:
Acceptance done 👍
